File: simpleadapter/src/main/java/com/github/deckyfx/simpleadapter/AdapterDataSet.java
package com.github.deckyfx.simpleadapter;

import android.support.annotation.NonNull;

import java.io.Serializable;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collection;
import java.util.Iterator;
import java.util.List;
import java.util.ListIterator;
import java.util.RandomAccess;

/**
 * Created by decky on 5/21/15.
 */


public class AdapterDataSet<E extends BaseItem> extends ArrayList<E> implements List<E>, RandomAccess, Cloneable, Serializable {
    public AdapterDataSet() {
        super();
    }

    public AdapterDataSet(Collection<? extends E> c) {
        super(c);
    }

    public AdapterDataSet<E> find(CharSequence constraint) {
        AdapterDataSet<E> results = new AdapterDataSet<E>();
        for (int i = 0; i < this.size(); i++) {
            if (this.get(i).testFilter(constraint)) {
                results.add(this.get(i));
            }
        }
        return results;
    }

    @Override
    public E get(int i) {
        return super.get(i);
    }

    @Override
    public int size() {
        return super.size();
    }
}

File: simpleadapter-kotlin/src/main/java/com/itp/android/simpleadapter_kotlin/AdapterDataSet.kt
package com.itp.android.simpleadapter_kotlin

import java.io.Serializable
import java.util.ArrayList
import java.util.RandomAccess

/**
 * Created by decky on 5/21/15.
 */


class AdapterDataSet<E : AdapterItem> : ArrayList<E>(), List<E>, RandomAccess, Cloneable, Serializable {
    fun find(constraint: CharSequence): AdapterDataSet<E> {
        val results = AdapterDataSet<E>()
        for (i in 0..this.size - 1) {
            if (this[i].testFilter(constraint)) {
                results.add(this[i])
            }
        }
        return results
    }

    override fun get(i: Int): E {
        return super.get(i)
    }

    override fun clone(): Any {
        return super<ArrayList>.clone();
    }
}
